module/zfs/dsl_dir.c

```diff
@@ -2280,6 +2288,66 @@ dsl_dir_remove_livelist(dsl_dir_t *dd, dmu_tx_t *tx, boolean_t total)
 	}
 }
 
+static int
+dsl_dir_activity_in_progress(dsl_dir_t *dd, dsl_dataset_t *ds,
+    zfs_wait_activity_t activity, boolean_t *in_progress)
+{
+	int error = 0;
+
+	ASSERT(MUTEX_HELD(&dd->dd_activity_lock));
+
+	switch (activity) {
+	case ZFS_WAIT_DELETEQ: {
+		objset_t *os;
+		error = dmu_objset_from_ds(ds, &os);
+		if (error != 0)
+			break;
+
+		uint64_t count, unlinked_obj;
+		error = zap_lookup(os, MASTER_NODE_OBJ, ZFS_UNLINKED_SET, 8, 1,
+		    &unlinked_obj);
+		if (error != 0) {
+			dsl_dataset_rele(ds, FTAG);
+			break;
+		}
+		error = zap_count(os, unlinked_obj, &count);
+
+		if (error == 0)
+			*in_progress = (count != 0);
+		break;
+	}
+	default:
+		panic("unrecognized value for activity %d", activity);
+	}
+
+	return (error);
+}
+
+int
+dsl_dir_wait(dsl_dir_t *dd, dsl_dataset_t *ds, zfs_wait_activity_t activity,
+    boolean_t *waited)
+{
+	int error = 0;
+	boolean_t in_progress;
+	mutex_enter(&dd->dd_activity_lock);
+	for (;;) {
+		error = dsl_dir_activity_in_progress(dd, ds, activity,
+		    &in_progress);
+		if (error != 0 || !in_progress)
+			break;
+
+		*waited = B_TRUE;
+
+		if (cv_wait_sig(&dd->dd_activity_cv, &dd->dd_activity_lock) ==
+		    0 || dd->dd_activity_cancelled) {
+			error = EINTR;
+			break;
+		}
```

Comment:
I see that we do something similar in spa_wait_common().  But are you sure that the return value of 0 indicates that a signal was received?  Looking at the implementation, it doesn't seem that way: 
```
int
__cv_wait_sig(kcondvar_t *cvp, kmutex_t *mp)
{
	cv_wait_common(cvp, mp, TASK_INTERRUPTIBLE, 0);

	return (signal_pending(current) ? 0 : 1);
}
```
Also see the definition of `issig()`, which does: `return (signal_pending(current));`

I think the typical way we do this in ZFS is to ignore the return value of `cv_wait_sig`, and when it returns we check `issig()` to see if we should abort with EINTR.

Reply:
Doesn't that do exactly the same thing? `signal_pending(current) ? 0  :  1` returns 0 if there's a signal  pending, so it returns 0 if `issig()` would return true ?

Reply:
I must have read that 3 times and couldn't get the logic straight in my head.  Yes you are right.